Python/pystate.c

```diff
@@ -1636,18 +1636,28 @@ PyThreadState_Clear(PyThreadState *tstate)
 
     int verbose = _PyInterpreterState_GetConfig(tstate->interp)->verbose;
 
-    if (verbose && tstate->current_frame != NULL) {
-        /* bpo-20526: After the main thread calls
-           _PyInterpreterState_SetFinalizing() in Py_FinalizeEx()
-           (or in Py_EndInterpreter() for subinterpreters),
-           threads must exit when trying to take the GIL.
-           If a thread exit in the middle of _PyEval_EvalFrameDefault(),
-           tstate->frame is not reset to its previous value.
-           It is more likely with daemon threads, but it can happen
-           with regular threads if threading._shutdown() fails
-           (ex: interrupted by CTRL+C). */
-        fprintf(stderr,
-          "PyThreadState_Clear: warning: thread still has a frame\n");
+    if (tstate->current_frame != NULL) {
+        _PyInterpreterFrame *frame = tstate->current_frame;
+        if (verbose) {
+            /* bpo-20526: After the main thread calls
+            _PyInterpreterState_SetFinalizing() in Py_FinalizeEx()
+            (or in Py_EndInterpreter() for subinterpreters),
+            threads must exit when trying to take the GIL.
+            If a thread exit in the middle of _PyEval_EvalFrameDefault(),
+            tstate->frame is not reset to its previous value.
+            It is more likely with daemon threads, but it can happen
+            with regular threads if threading._shutdown() fails
+            (ex: interrupted by CTRL+C). */
+            fprintf(stderr,
+            "PyThreadState_Clear: warning: thread still has a frame\n");
+        }
+        do {
+            if (frame->owner == FRAME_OWNED_BY_GENERATOR) {
+                PyGenObject *gen = _PyGen_GetGeneratorFromFrame(frame);
+                gen->gi_frame_state = FRAME_ZOMBIE;
+            }
+            frame = frame->previous;
+        } while (frame != NULL);
```

Comment:
It's not sufficient to mark generators as zombies in `PyThreadState_Clear()`. We call `PyThreadState_Clear()` one at a time on each non-main thread:

* At that point, they are already unlinked from the interpreter's list of `PyThreadStates`
* The `PyThreadState_Clear()` calls makes escaping calls (`Py_CLEAR()`) that can trigger the GC

So the GC may still see generators that are marked as "running", but aren't in any accessible `PyThreadState`'s frame stack.